evm/utils/blobs.py

```diff
@@ -60,5 +65,63 @@ def calc_chunk_root(collation_body: bytes) -> bytes:
             COLLATION_SIZE
         ))
 
-    chunks = chunk_iterator(collation_body)
+    chunks = iterate_chunks(collation_body)
     return calc_merkle_root(chunks)
+
+
+def check_body_size(body):
+    if len(body) > COLLATION_SIZE:
+        raise ValueError("Collation body exceeds maximum allowed size")
+    return body
+
+
+@apply_to_return_value(check_body_size)
+@apply_to_return_value(lambda v: zpad_right(v, COLLATION_SIZE))
+@apply_to_return_value(b"".join)
+def serialize_blobs(blobs: Iterable[bytes]) -> bytes:
```

Comment:
@jannikluhn You probably annotated the return value as `bytes` because that is the type that the caller of `serialize_blobs` sees. At least, this confused me a lot when I first came across something like this because something inside me naturally wanted to annotate the return type as the type it is **after all decorators are applied**. However, you do need to make the return type `Iterator[bytes]` and still on the call site `mypy` will infer the correct type (e.g. `bytes` in this case) based on the applied decorators.

Reply:
It confused me indeed, only realized this when mypy was failing, but in hindsight it makes sense.

> and still on the call site mypy will infer the correct type (e.g. bytes in this case) based on the applied decorators.

Is this true even with the unannotated lambda function in the decorator or should I replace it with a `def` function (or probably better `partial(zpad_right, to_size=COLLATION_SIZE)`)?

Reply:
>It confused me indeed, only realized this when mypy was failing, but in hindsight it makes sense

You have no idea how much better I feel now, knowing I wasn't the only one to think like that ;)

>Is this true even with the unannotated lambda function in the decorator

In general, I do believe that the type inference would work out the lamda just fine. However, in our case the type system will still fall back to `Any` (yeah, bummer!) which I *believe* can only be fixed through a stronger [`apply_to_return_value`](https://github.com/ethereum/eth-utils/blob/a1a2ee20988f56f6f99e31e2eff4b98e10c35814/eth_utils/functional.py#L18-L25) decorator.

I think with the current decorator as it stands, whatever you apply it on will make the type system blind to figure out the proper return type.

Reply:
Maybe we can do something about it https://github.com/ethereum/eth-utils/pull/88